File: solutions/golang/hotelmanagement/reservation.go
package hotelmanagement

import (
	"fmt"
	"sync"
	"time"
)

type Reservation struct {
	ID           string
	Guest        *Guest
	Room         *Room
	CheckInDate  time.Time
	CheckOutDate time.Time
	Status       ReservationStatus
	mu           sync.RWMutex
}

func NewReservation(id string, guest *Guest, room *Room, checkInDate, checkOutDate time.Time) *Reservation {
	return &Reservation{
		ID:           id,
		Guest:        guest,
		Room:         room,
		CheckInDate:  checkInDate,
		CheckOutDate: checkOutDate,
		Status:       ReservationStatusConfirmed,
	}
}

func (r *Reservation) Cancel() error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.Status != ReservationStatusConfirmed {
		return fmt.Errorf("reservation is not confirmed")
	}

	r.Status = ReservationStatusCancelled
	return r.Room.CheckOut()
}

File: solutions/csharp/hotelmanagementsystem/Reservation.cs
using System;

namespace HotelManagement
{
    public class Reservation
    {
        public string Id { get; }
        public Guest Guest { get; }
        public Room Room { get; }
        public DateTime CheckInDate { get; }
        public DateTime CheckOutDate { get; }
        public ReservationStatus Status { get; private set; }

        public Reservation(string id, Guest guest, Room room, DateTime checkInDate, DateTime checkOutDate)
        {
            Id = id;
            Guest = guest;
            Room = room;
            CheckInDate = checkInDate;
            CheckOutDate = checkOutDate;
            Status = ReservationStatus.CONFIRMED;
        }

        public void Cancel()
        {
            if (Status == ReservationStatus.CONFIRMED)
            {
                Status = ReservationStatus.CANCELLED;
                Room.CheckOut();
            }
            else
            {
                throw new InvalidOperationException("Reservation is not confirmed.");
            }
        }
    }
}